### PR TITLE
HDDS-7713. listKeys with prefix path as key for FSO buckets is redundant

### PR DESCRIPTION
**Pull request by Tejaskriya**
## What changes were proposed in this pull request?

The redundant listing of keys was originating from the client module (it was visible from both `aws s3 ls` and `ozone sh key list` interfaces). The change proposed is to not add the prefix to the result list if the prefix is for a key and not a directory.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7713

## How was this patch tested?

Tested manually using docker. 
Through S3:
```
$ aws s3 ls --endpoint http://localhost:9878 s3://buck1/k1  
2023-08-10 13:37:14          3 k1
```
Through ozone sh:
```
$ ozone sh key list s3v/buck1 --prefix k1
[ {
  "volumeName" : "s3v",
  "bucketName" : "buck1",
  "name" : "k1",
  "dataSize" : 3,
  "creationTime" : "2023-08-10T08:07:13.130Z",
  "modificationTime" : "2023-08-10T08:07:14.138Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { },
  "file" : true
} ]

```

Also added a test in testListKeysAtDifferentLevels() of TestObjectStoreWithFSO.

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestObjectStoreWithFSO.java [hunk 1]
```diff
@@ -473,6 +473,13 @@ public void testListKeysAtDifferentLevels() throws Exception {
         ozoneBucket.listKeys("a/b3/e3", "a/b3/e3/e31.tx");
     expectedKeys = new LinkedList<>();
     checkKeyList(ozoneKeyIterator, expectedKeys);
+
+    // Key level, prefix=key case
+    ozoneKeyIterator =
+        ozoneBucket.listKeys("a/b1/c1/c1.tx");
+    expectedKeys = new LinkedList<>();
+    expectedKeys.add("a/b1/c1/c1.tx");
+    checkKeyList(ozoneKeyIterator, expectedKeys);
```
Comment (sadanand48) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestObjectStoreWithFSO.java:482]: maybe add one case for a list key on a dir prefix :
ozoneBucket.listKeys("a/b1/c1/") and one for a top level key in a bucket

**Comment by whbing** [comment]
LGTM Overall. Thanks ! Here's a minor suggestion from my perspective. 
We can add test in `TestListKeysWithFSO` to compare the listKeys results of `legacyBucket` and `fsoBucket`. I've written a demo test in `TestListKeysWithFSO`:
```java
private static void buildNameSpaceTree(OzoneBucket ozoneBucket)
      throws Exception {
    LinkedList<String> keys = new LinkedList<>();
    // ignore other codes    
    keys.add("/a1/b3/e3/e31.tx");
    keys.add("/a1/b3/e3/e31.tx1");
   // ignore other codes
    createKeys(ozoneBucket, keys);
  }

  @Test
  public void testListKeysWithValidStartKey() throws Exception {
    // case-10: keyPrefix corresponds an exist file and other file's prefix
    expectedKeys =
        getExpectedKeyList("a1/b3/e3/e31.tx", "", legacyOzoneBucket);
    checkKeyList("a1/b3/e3/e31.tx", "", expectedKeys, fsoOzoneBucket);
  }
```
Result will be `Actual   :[a1/b3/e3/e31.tx, a1/b3/e3/e31.tx] ` without PR  and `Actual   :[a1/b3/e3/e31.tx]` with this PR.
Since the result from `fsoBucket` is not consistent with `legacybucket` in the above case (In fact, I'v  noticed this duplication problem, as well as other omission problem in fsoBucket, see issue [HDDS-9065](https://issues.apache.org/jira/browse/HDDS-9065)), we can first modify the test to:
```java
  @Test
  public void testListKeysWithValidStartKey() throws Exception {
    // case-10: keyPrefix corresponds an exist file and other file's prefix
    expectedKeys = newLinkedList(Arrays.asList("a1/b3/e3/e31.tx"));
    checkKeyList("a1/b3/e3/e31.tx", "", expectedKeys, fsoOzoneBucket);
  }
```

**Comment by whbing** [comment]
NIT: title can be specified  FSO

**Comment by Tejaskriya** [comment]
> ```java
> // case-10: keyPrefix corresponds an exist file
>     expectedKeys =
>         getExpectedKeyList("a1/b3/e3/e31.tx", "", legacyOzoneBucket);
>     checkKeyList("a1/b3/e3/e31.tx", "", expectedKeys, fsoOzoneBucket);
> ```

Thank you for the suggestion, I have added the test case you've specified in  `TestListKeysWithFSO`

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestObjectStoreWithFSO.java (hunk 1 above)
Comment (Tejaskriya) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestObjectStoreWithFSO.java:487]: Thank you for the review! Directory case was covered in one of the existing cases-
```
ozoneKeyIterator =
        ozoneBucket.listKeys("a/b2/d1", null);
    expectedKeys = new LinkedList<>();
    expectedKeys.add("a/b2/d1/");
    expectedKeys.add("a/b2/d1/d11.tx");
    checkKeyList(ozoneKeyIterator, expectedKeys);
```
So instead, I have added a test case with trailing slash for dir prefix, and another one for key directly under the bucket.

**Comment by whbing** [comment]
+1, once ci passed.

**Review by sadanand48 — APPROVED** [review]
LGTM

**Review by aryangupta1998 — APPROVED** [review]
LGTM, +1

**Comment by sadanand48** [comment]
Thanks @Tejaskriya for the patch, @whbing @aryangupta1998 for the review.

**Pull request merged**